slyd/slyd/resources/projects.py

```diff
@@ -1,144 +1,230 @@
-            for type_, path, old_path in manager.pm._changed_files(project_id)]
-    }
-    return data
-
-
-def merge(manager, attributes=None):
-    if not hasattr(manager.pm, 'publish_project'):
-        raise NotFound(NOT_AVAILABLE_ERROR)
-    project_id = manager.project_name
-    if not status(manager):
-        raise BadRequest('The project is up to date')
-    publish_status = json.loads(
-        manager.pm.publish_project(project_id, attributes.get('force', False)))
-    if publish_status['status'] == 'conflict':
-        raise BaseError(409, 'A conflict has occurred in this project',
-                        'You must resolve the conflict for the project to be'
-                        ' successfully published')
-    data = get_project(manager)
-    data['meta'] = manager.pm._schedule_data(project_id=project_id)
-    return data
-
-
-def reset(manager, attributes=None):
-    if not hasattr(manager.pm, 'discard_changes'):
-        raise NotFound(NOT_AVAILABLE_ERROR)
-    project_id = manager.project_name
-    if not status(manager):
-        raise BadRequest('There are no changes to discard')
-    manager.pm.discard_changes(project_id)
-    return get_project(manager)
-
-
-class ProjectsManagerFileResponse(BaseApiResponse):
-    def __init__(self, data, command, project_manager):
-        super(ProjectsManagerFileResponse, self).__init__(data)
-        self.command = command
-        self.project_manager = project_manager
-
-    def format_response(self, request):
-        return self.project_manager._render_file(request, self.command,
-                                                 self.data)
-
-
-def download(manager, spider_id=None, attributes=None):
-    project_id = manager.project_name
-    spider_ids = [spider_id] if spider_id is not None else '*'
-    command = {
-        'cmd': 'download',
-        'args': [project_id, spider_ids]
-    }
-    fmt = attributes['arguments'].get('format', ['spec'])[0]
-    file_content = manager.pm.download_project(project_id, spider_ids, fmt=fmt)
-    return ProjectsManagerFileResponse(file_content, command, manager.pm)
+class ProjectDownloadMixin(object):
+    @classmethod
+    def get_resources(cls):
+        for resouce in super(ProjectDownloadMixin, cls).get_resources():
+            yield resouce
+        yield 'get', cls.download_path
+
+    def get_handler(self):
+        if self.route_path == self.download_path:
+            return self.download
+        return super(ProjectDownloadMixin, self).get_handler()
+
+    def download(self):
+        project_manager = self.project_manager
+        project_id = self.args.get('project_id')
+        spider_id = self.args.get('spider_id', None)
+        spider_ids = [spider_id] if spider_id is not None else '*'
+        fmt = self.query.get('format', ['spec'])[0]
+        return ProjectDownloadResponse(
+            project_id, spider_ids, fmt, project_manager)
+
+
+class ProjectDataMixin(object):
```

Comment:
I think it makes sense to have mixins for the different CRUD verbs for routes as these mixins need to have available much of the state of the object. However, in this case, the `ProjectDataMixin` only accesses two instance variables `self.request.auth_info` and `self.project_manager.all_projects`. We can have a class which accepts the `auth_info` and `projects` variables to its constructor and use composition instead of inheritance (mixin) for it to be used in the `ProjectRoute` and `SpiderRoute`.